ceno_host/src/lib.rs

```diff
@@ -0,0 +1,87 @@
+use std::iter::zip;
+
+use anyhow::Result;
+use ceno_emul::{IterAddresses, Platform, VMState, host_utils::read_all_messages};
+use itertools::izip;
+use rkyv::{
+    Serialize, api::high::HighSerializer, rancor::Error, ser::allocator::ArenaHandle, to_bytes,
+    util::AlignedVec,
+};
+
+// We want to get access to the default value of `AlignedVec::ALIGNMENT`, and using it directly like this
+//   pub const RKVY_ALIGNMENT: usize = rkyv::util::AlignedVec::ALIGNMENT;
+// doesn't work:
+pub const RKYV_ALIGNMENT: usize = {
+    type AlignedVec = rkyv::util::AlignedVec;
+    AlignedVec::ALIGNMENT
+};
+
+#[derive(Default)]
+pub struct CenoStdin {
+    pub items: Vec<AlignedVec>,
+}
+
+impl CenoStdin {
+    pub fn write_slice(&mut self, bytes: AlignedVec) {
+        self.items.push(bytes);
+    }
+
+    pub fn write(
+        &mut self,
+        item: &impl for<'a> Serialize<HighSerializer<AlignedVec, ArenaHandle<'a>, Error>>,
+    ) -> Result<(), Error> {
+        to_bytes::<Error>(item).map(|bytes| self.write_slice(bytes))
+    }
+
+    pub fn finalise(&self) -> Vec<u32> {
+        let initial_offset = (size_of::<u32>() * self.items.len()).next_multiple_of(RKYV_ALIGNMENT);
+        let offsets: Vec<u32> = self
+            .items
+            .iter()
+            .scan(initial_offset, |acc, bytes| {
+                let output = (*acc + bytes.len()) as u32;
```

Comment:
why you don't increase the `bytes.len()` up to multiple of `RKYV_ALIGNMENT`?

Reply:
Do you mean extending the buffer itself?  We could do that, but then we'd need to keep track of the original offsets (= lengths of the buffers) ourselves.

That's because rkyv wants its buffers to start at a 16 alignment, but it also wants the buffers to end exactly how they come out of its own `.to_bytes` function.  You can't just arbitrarily pad them to the right.

You can however pad them to the left just fine, as long as you keep the 16 byte alignment in mind.

Reply:
My question is that in Line 43, the `acc` increases by the byte lens rounded to multiples of `RKYV_ALIGNMENT`. It seems to contradict the `output`, which calculates the offsets, right?